elastic/kibana
# Convert hosts flyout contents to asset detail embeddable

Closes #156687

## Summary

This PR adds asset details embeddable. The idea of extracting the existing host details content to an embeddable is to make it reusable in different places. It can be used as a full-page view or as a flyout controlled by the `showInFlyout` prop. 

## Usage

### `AssetDetails` embeddable

The embeddable can be customized - the user of the embeddable can decide if it should be shown in a flyout or as a full-page view, which tabs and links should be displayed, and if the state should be persisted (passing the state and the setter). These configurations are supported as props:
- node: the node whose details should be shown
- closeFlyout: function to determine when to close the flyout
- onTabClick: function to decide which tab content should be visible
- currentTimeRange- the time range `from` and `to` used to get the content
- [Optional] If the state should be persisted in  the URL like in the host details flyout
    - hostFlyoutOpen
    - setHostFlyoutOpen
- [Optional] showActionsColumn: should the actions column inside the metadata table be visible (not displayed by default)
- [Optional] showInFlyout: should the content be shown in a flyout (full-page by default)
- [Optional] renderedTabsSet: function to handle which tab should be shown externally
- tabs: array of the tabs that should be displayed
- [Optional] links: array of the link names that should be included (Supported link currently: ['apmServices', 'uptime']) - no links are displayed by default 
- nodeType: Currently only `host` is supported

### `Metadata` embeddable

The metadata tab content can be used separately from the other content as embeddable. I can be configured to show actions and to use an external search state. Props supported: 
  currentTimeRange: MetricsTimeInput;
  node: the node whose details should be shown
  nodeType: Currently only `host` is supported
- [Optional] showActionsColumn: should the actions column inside the metadata table be visible (not displayed by default);
- [Optional] persistMetadataSearchToUrlState If the state of the metadata search should be persisted in the URL 
    - metadataSearchUrlState: this string is used to get the search term and prefill it
    - setMetadataSearchUrlState: function to set the search term

## 🐾 Next steps

We are currently still using the old inventory component and that leads to some duplicated code - I tried to move as much as possible but I will suggest removing the components from the inventory node details once the new embeddable is used there. Ideally, all the tabs will be removed from the inventory components (+ translations) and added to the embeddable (in a separate issue)

The telemetry should be also implemented - once we implement the other tabs and use it in different apps/pages we can implement custom test ids based on where the embeddable is used - to keep the test working without changing the ids there I am keeping the host test ids but we need to decide how we want to build the test ids (for example if we use it in APM should the id be `apm-flyout-metadata-remove-filter` instead of `hostsView-flyout-metadata-remove-filter` or we want to pass totally different ids - which is more complicated.


## Testing
1. Storybook
   - run ```yarn storybook infra```
   - The Host Details View folder should be visible - **renamed to Asset Details View**
   
<img width="1935" alt="image" src="https://github.com/elastic/kibana/assets/14139027/b99bf4e0-4380-4d93-9d4b-7834285c6979">

   - Check the content: A quick guide -
<img width="2551" alt="Screenshot 2023-05-23 at 15 47 48" src="https://github.com/elastic/kibana/assets/14139027/d06585e2-ae32-48ee-814a-f4bd5d96fb1a">

 2. Hosts view
   - Open hosts view and open a host details flyout
     - It should work as before - all tabs and links should be functional and the state should be persisted in the URL if it's opened. 
     - The filter button should be centered inside the actions column 
<img width="1930" alt="image" src="https://github.com/elastic/kibana/assets/14139027/bc09601c-40ff-4d3c-893e-d695f7b6737e">
